src/python/pants/backend/python/rules/python_fmt.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import os
+
+from pants.backend.python.rules.create_requirements_pex import (
+  RequirementsPex,
+  RequirementsPexRequest,
+)
+from pants.backend.python.subsystems.black import Black
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.subprocess_environment import SubprocessEncodingEnvironment
+from pants.engine.fs import Digest, DirectoriesToMerge, PathGlobs, Snapshot
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.structs import (
+  PantsPluginAdaptor,
+  PythonAppAdaptor,
+  PythonBinaryAdaptor,
+  PythonTargetAdaptor,
+  PythonTestsAdaptor,
+)
+from pants.engine.rules import UnionRule, optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.rules.core.fmt import FmtResult, FmtTarget
+from pants.util.objects import datatype
+
+# Note: this is a workaround until https://github.com/pantsbuild/pants/issues/8343 is addressed
+# We have to write this type which basically represents a union of all various kinds of targets
+# containing python files so we can have one single type used as an input in the run_black rule.
+class FormattablePythonTarget(datatype(['target'])):
+  pass
+
+
+@rule
+def run_black(
+  wrapped_target: FormattablePythonTarget,
+  black: Black,
+  python_setup: PythonSetup,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment,
+  ) -> FmtResult:
+  config_path = black.get_options().config
+  config_snapshot = yield Get(Snapshot, PathGlobs, PathGlobs(include=(config_path,)))
+
+  resolved_requirements_pex = yield Get(
+    RequirementsPex, RequirementsPexRequest(
+      output_filename="black.pex",
+      requirements=tuple(black.get_requirement_specs()),
+      interpreter_constraints=tuple(black.default_interpreter_constraints),
+      entry_point=black.get_entry_point(),
+    )
+  )
+  target = wrapped_target.target
+  sources_digest = target.sources.snapshot.directory_digest
+
+  all_input_digests = [
+    sources_digest,
+    resolved_requirements_pex.directory_digest,
+    config_snapshot.directory_digest,
+  ]
+  merged_input_files = yield Get(
+    Digest,
+    DirectoriesToMerge,
+    DirectoriesToMerge(directories=tuple(all_input_digests)),
+  )
+
+  # The exclude option from Black only works on recursive invocations,
+  # so call black with the directories in which the files are present
+  # and passing the full file names with the include option
+  dirs = set()
```

Comment:
Empty collections like this should be typed:

```python
from typing import Set

dirs: Set[Path] = set()
```

(Path is if you take the below recommendation. Otherwise, it's `Set[str]`.